source/common/quic/envoy_quic_session_cache.cc

```diff
@@ -0,0 +1,168 @@
+  cache_.erase(it);
+  createAndInsertEntry(server_id, std::move(session), params, application_state);
+}
+
+std::unique_ptr<quic::QuicResumptionState>
+EnvoyQuicSessionCache::Lookup(const quic::QuicServerId& server_id, const SSL_CTX* /*ctx*/) {
+  auto it = cache_.find(server_id);
+  if (it == cache_.end()) {
+    return nullptr;
+  }
+  Entry& entry = it->second;
+  const SystemTime system_time = time_source_.systemTime();
+  if (!isSessionValid(entry.peekSession(), system_time)) {
+    cache_.erase(it);
+    return nullptr;
+  }
+  auto state = std::make_unique<quic::QuicResumptionState>();
+  state->tls_session = entry.popSession();
+  if (entry.params != nullptr) {
+    state->transport_params = std::make_unique<quic::TransportParameters>(*entry.params);
+  }
+  if (entry.application_state != nullptr) {
+    state->application_state = std::make_unique<quic::ApplicationState>(*entry.application_state);
+  }
+  return state;
+}
+
+void EnvoyQuicSessionCache::ClearEarlyData(const quic::QuicServerId& server_id) {
+  auto it = cache_.find(server_id);
+  if (it == cache_.end()) {
+    return;
+  }
+  for (bssl::UniquePtr<SSL_SESSION>& session : it->second.sessions) {
+    if (session != nullptr) {
+      session.reset(SSL_SESSION_copy_without_early_data(session.get()));
+    }
+  }
+}
+
+void EnvoyQuicSessionCache::prune() {
+  quic::QuicServerId oldest_id;
+  uint64_t oldest_expiration = std::numeric_limits<uint64_t>::max();
+  const SystemTime system_time = time_source_.systemTime();
+  auto it = cache_.begin();
+  while (it != cache_.end()) {
+    Entry& entry = it->second;
+    SSL_SESSION* session = entry.peekSession();
+    if (!isSessionValid(session, system_time)) {
+      it = cache_.erase(it);
+    } else {
+      if (cache_.size() >= kMaxSessionCacheEntries) {
+        // Only save oldest if we are at the size limit.
+        const uint64_t session_expiration =
+            SSL_SESSION_get_time(session) + SSL_SESSION_get_timeout(session);
+        if (session_expiration < oldest_expiration) {
+          oldest_expiration = session_expiration;
+          oldest_id = it->first;
+        }
+      }
+      ++it;
+    }
+  }
+  if (cache_.size() >= kMaxSessionCacheEntries) {
+    cache_.erase(oldest_id);
+  }
+}
+
+void EnvoyQuicSessionCache::createAndInsertEntry(const quic::QuicServerId& server_id,
+                                                 bssl::UniquePtr<SSL_SESSION> session,
+                                                 const quic::TransportParameters& params,
+                                                 const quic::ApplicationState* application_state) {
+  prune();
+  ASSERT(cache_.size() < kMaxSessionCacheEntries);
+  Entry entry;
+  entry.pushSession(std::move(session));
+  entry.params = std::make_unique<quic::TransportParameters>(params);
+  if (application_state != nullptr) {
+    entry.application_state = std::make_unique<quic::ApplicationState>(*application_state);
+  }
```

Comment:
nits: consider wrapping these into Entry constructor.

Reply:
Given that this is a struct which needs a default constructor, I think this is overall simpler?